File: app/src/main/java/com/trikset/gamepad/MainActivity.java
package com.trikset.gamepad;

import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.ActivityInfo;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.util.Log;
import android.view.Gravity;
import android.view.HapticFeedbackConstants;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewGroup.LayoutParams;
import android.view.animation.AlphaAnimation;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.view.MenuItemCompat;
import androidx.core.view.WindowCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.core.view.WindowInsetsControllerCompat;
import com.demo.mjpeg.MjpegView;
import java.net.MalformedURLException;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.URL;
import java.util.Locale;
import java.util.Objects;

public class MainActivity extends AppCompatActivity implements SensorEventListener {
  static final String TAG = "MainActivity";
  private HideRunnable mHideRunnable;
  private SensorManager mSensorManager;
  private int mAngle; // -100%
  // ...
  // +100%
  private boolean mWheelEnabled;
  private SenderService mSender;
  private int mWheelStep = 7;
  @Nullable private MjpegView mVideo;
  @Nullable private URL mVideoURL;
  @Nullable private SharedPreferences.OnSharedPreferenceChangeListener mSharedPreferencesListener;

  // @SuppressWarnings("deprecation")
  // @TargetApi(16)
  private void createPad(int id, String strId) {
    final SquareTouchPadLayout pad = findViewById(id);
    if (pad != null) {
      pad.setPadName("pad " + strId);
      pad.setSender(getSenderService());
    }
    // if (android.os.Build.VERSION.SDK_INT >= 16) {
    // pad.setBackground(image);
    // } else {
    // pad.setBackgroundDrawable(image);
    // }
  }

  @Override
  public void onAccuracyChanged(final Sensor arg0, final int arg1) {
    // TODO Auto-generated method stub

  }

  @Override
  protected void onCreate(final Bundle savedInstanceState) {
    setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE);
    // Edge-to-edge (required on API 35+; enforced for targetSdk 36): draw
    // behind the system bars instead of using the removed FLAG_FULLSCREEN.
    // The gamepad UI hides the bars via setSystemUiVisibility(false).
    WindowCompat.setDecorFitsSystemWindows(getWindow(), false);
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);
    setHideRunnable(new HideRunnable());
    setSystemUiVisibility(false);
    {
      ActionBar a = getSupportActionBar();
      if (a != null) {

        a.setDisplayShowHomeEnabled(true);
        a.setDisplayUseLogoEnabled(false);
        a.setLogo(R.drawable.trik_gamepad_logo_512x512);

        a.setDisplayShowTitleEnabled(true);
      }
    }

    setSenderService(new SenderService());
    mSensorManager = (SensorManager) getSystemService(SENSOR_SERVICE);

    mVideo = findViewById(R.id.video);

    recreateMagicButtons(5);

    {
      getSenderService().setOnDisconnectedListener(reason -> toast("Disconnected." + reason));
      getSenderService().setShowTextCallback(this::toast);
    }

    {
      final Button btnSettings = findViewById(R.id.btnSettings);
      if (btnSettings != null) {
        btnSettings.setOnClickListener(
            v -> {
              ActionBar a = getSupportActionBar();
              if (a != null) {
                setSystemUiVisibility(!a.isShowing());
              }
            });
      }
    }

    {
      final View controlsOverlay = findViewById(R.id.controlsOverlay);
      if (controlsOverlay != null) {
        controlsOverlay.bringToFront();
      }
    }

    {
      createPad(R.id.leftPad, "1");
      createPad(R.id.rightPad, "2");
    }

    {
      final SharedPreferences prefs =
          PreferenceManager.getDefaultSharedPreferences(getBaseContext());
      mSharedPreferencesListener =
          new SharedPreferences.OnSharedPreferenceChangeListener() {
            private float mPrevAlpha;

            @Override
            public void onSharedPreferenceChanged(
                @NonNull final SharedPreferences sharedPreferences, final String key) {
              final String addr =
                  sharedPreferences.getString(SettingsFragment.SK_HOST_ADDRESS, "192.168.77.1");
              int portNumber = 4444;
              final String portStr =
                  sharedPreferences.getString(SettingsFragment.SK_HOST_PORT, "4444");
              try {
                portNumber = Integer.parseInt(portStr);
              } catch (@NonNull final NumberFormatException e) {
                toast("Port number '" + portStr + "' is incorrect.");
              }
              final String oldAddr = getSenderService().getHostAddr();
              getSenderService().setTarget(addr, portNumber);

              {
                ActionBar a = getSupportActionBar();

                if (a != null) {
                  a.setTitle(addr);
                } else {
                  toast("Can not change title, not a problem");
                }
              }

              if (!addr.equalsIgnoreCase(oldAddr)) {

                // update video stream URI when target addr changed
                sharedPreferences
                    .edit()
                    .putString(
                        SettingsFragment.SK_VIDEO_URI,
                        "http://" + addr.trim() + ":8080/?action=stream")
                    .apply();
              }

              {
                final int defAlpha = 100;
                int padsAlpha = defAlpha;

                try {
                  padsAlpha =
                      Integer.parseInt(
                          sharedPreferences.getString(
                              SettingsFragment.SK_SHOW_PADS, String.valueOf(defAlpha)));
                } catch (NumberFormatException nfe) {
                  // unchanged
                }

                final float alpha = Math.max(0, Math.min(255, padsAlpha)) / 255.0f;
                AlphaAnimation alphaUp = new AlphaAnimation(mPrevAlpha, alpha);
                mPrevAlpha = alpha;
                alphaUp.setFillAfter(true);
                alphaUp.setDuration(2000);
                final View co = findViewById(R.id.controlsOverlay);
                if (co != null) {
                  co.startAnimation(alphaUp);
                }
                final View btns = findViewById(R.id.buttons);
                if (btns != null) {
                  btns.startAnimation(alphaUp);
                }
              }

              {
                // "http://trackfield.webcam.oregonstate.edu/axis-cgi/mjpg/video.cgi?resolution=320x240";

                String videoStreamURI =
                    sharedPreferences.getString(
                        SettingsFragment.SK_VIDEO_URI, "http://" + addr + ":8080/?action=stream");

                // --no-sout-audio --sout
                // "#transcode{width=320,height=240,vcodec=mp2v,fps=20}:"
                // "rtp{ttl=5,sdp=rtsp://:8889/s}"
                // works only with vcodec=mp4v without audio :(

                // http://developer.android.com/reference/android/media/MediaPlayer.html
                // http://developer.android.com/guide/appendix/media-formats.html

                try {
                  mVideoURL = videoStreamURI.isEmpty() ? null : new URI(videoStreamURI).toURL();
                } catch (URISyntaxException | MalformedURLException e) {
                  toast("Illegal video stream URL");
                  Log.e(TAG, "onSharedPreferenceChanged: ", e);
                  mVideoURL = null;
                }
              }

              {
                mWheelStep =
                    Objects.requireNonNull(
                        Integer.getInteger(
                            sharedPreferences.getString(
                                SettingsFragment.SK_WHEEL_STEP, String.valueOf(mWheelStep)),
                            mWheelStep));
                mWheelStep = Math.max(1, Math.min(100, mWheelStep));
              }

              {
                try {
                  final int timeout =
                      Integer.parseInt(
                          sharedPreferences.getString(
                              SettingsFragment.SK_KEEPALIVE,
                              Integer.toString(SenderService.DEFAULT_KEEPALIVE)));
                  if (timeout < SenderService.MINIMAL_KEEPALIVE) {
                    toast(
                        String.format(
                            Locale.US,
                            getString(R.string.keepalive_must_be_not_less),
                            SenderService.MINIMAL_KEEPALIVE));

                    sharedPreferences
                        .edit()
                        .putString(
                            SettingsFragment.SK_KEEPALIVE,
                            Integer.toString(getSenderService().getKeepaliveTimeout()))
                        .apply();
                  } else {
                    getSenderService().setKeepaliveTimeout(timeout);
                  }
                } catch (NumberFormatException e) {
                  toast(getString(R.string.keepalive_must_be_positive_decimal));

                  sharedPreferences
                      .edit()
                      .putString(
                          SettingsFragment.SK_KEEPALIVE,
                          Integer.toString(getSenderService().getKeepaliveTimeout()))
                      .apply();
                }
              }
            }
          };
      mSharedPreferencesListener.onSharedPreferenceChanged(prefs, SettingsFragment.SK_HOST_ADDRESS);
      prefs.registerOnSharedPreferenceChangeListener(mSharedPreferencesListener);
    }
  }

  @Override
  public boolean onCreateOptionsMenu(@NonNull Menu menu) {
    MenuInflater inflater = getMenuInflater();
    inflater.inflate(R.menu.menu, menu);

    // TODO: remove this hack
    final CheckBox w = (CheckBox) MenuItemCompat.getActionView(menu.findItem(R.id.wheel));
    w.setText(getResources().getString(R.string.menu_wheel));

    return true;
  }

  @Override
  public boolean onOptionsItemSelected(@NonNull MenuItem item) {
    if (item.getItemId() == R.id.settings) {
      final Intent settings = new Intent(MainActivity.this, SettingsActivity.class);
      startActivity(settings);
      return true;
    } else if (item.getItemId() == R.id.wheel) {
      mWheelEnabled = !mWheelEnabled;
      item.setChecked(mWheelEnabled);
      return true;
    } else {
      return super.onOptionsItemSelected(item);
    }
  }

  @Override
  protected void onPause() {
    mSensorManager.unregisterListener(this);
    getSenderService().disconnect("Inactive gamepad");
    if (mVideo != null) {
      mVideo.stopPlayback();
      mVideo.setOnStreamErrorListener(null);
    }
    super.onPause();
  }

  @Override
  protected void onResume() {
    super.onResume();

    if (mVideo != null) {
      // Reconnect-on-error: the render thread reports a dead stream and we
      // drop the HTTP connection and restart it. No forced periodic restart —
      // the stream only restarts when it actually breaks (see .PLAN.md R12).
      mVideo.setOnStreamErrorListener(this::restartVideoStream);
      restartVideoStream();
    }

    mSensorManager.registerListener(
        this, mSensorManager.getDefaultSensor(Sensor.TYPE_ALL), SensorManager.SENSOR_DELAY_NORMAL);
  }

  private void restartVideoStream() {
    // The error listener may fire from the render thread; always hop to the
    // main thread before touching the view hierarchy / launching an AsyncTask.
    runOnUiThread(
        () -> {
          if (mVideo == null) {
            return;
          }
          new StartReadMjpegAsync(mVideo).execute(mVideoURL);
        });
  }

  @Override
  public void onSensorChanged(@NonNull final SensorEvent event) {
    if (event.sensor.getType() == Sensor.TYPE_ACCELEROMETER) {
      if (!mWheelEnabled) {
        return;
      }
      processSensor(event.values);
    } else {
      Log.i("Sensor", String.valueOf(event.sensor.getType()));
    }
  }

  private void processSensor(final float[] values) {
    final double WHEEL_BOOSTER_MULTIPLIER = 1.5;
    final double x = values[0];
    final double y = values[1];
    if (x < 1e-6) {
      return;
    }

    int angle = (int) (200 * WHEEL_BOOSTER_MULTIPLIER * Math.atan2(y, x) / Math.PI);

    if (Math.abs(angle) < 10) {
      angle = 0;
    } else if (angle > 100) {
      angle = 100;
    } else if (angle < -100) {
      angle = -100;
    }

    if (Math.abs(mAngle - angle) < mWheelStep) {
      return;
    }

    mAngle = angle;

    getSenderService().send("wheel " + mAngle);
  }

  private void recreateMagicButtons(final int count) {
    final ViewGroup buttonsView = findViewById(R.id.buttons);
    if (buttonsView == null) {
      return;
    }
    buttonsView.removeAllViews();
    for (int num = 1; num <= count; ++num) {
      final Button btn = new Button(MainActivity.this);
      btn.setHapticFeedbackEnabled(true);
      btn.setGravity(Gravity.CENTER);
      btn.setLayoutParams(new LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT));
      final String name = String.valueOf(num);
      btn.setText(name);
      btn.setBackgroundResource(R.drawable.button_shape);

      btn.setOnClickListener(
          arg0 -> {
            SenderService sender = getSenderService();
            if (sender != null) {
              sender.send("btn " + name + " down"); // TODO: "up" via
              // TouchListner
              btn.performHapticFeedback(
                  HapticFeedbackConstants.LONG_PRESS,
                  HapticFeedbackConstants.FLAG_IGNORE_GLOBAL_SETTING);
            }
          });
      buttonsView.addView(btn);
    }
  }

  // Hides/shows the system bars. Uses the modern WindowInsetsControllerCompat
  // (available back to API 21 via core-ktx) instead of the deprecated
  // View.SYSTEM_UI_FLAG_* set, which leaves the window without focus on API 36
  // where edge-to-edge is enforced.
  private void setSystemUiVisibility(boolean show) {
    final View mainView = findViewById(R.id.main);
    if (mainView == null) {
      return;
    }

    WindowInsetsControllerCompat controller =
        WindowCompat.getInsetsController(getWindow(), mainView);
    if (controller == null) {
      return;
    }

    if (show) {
      controller.show(WindowInsetsCompat.Type.systemBars());
      // The action bar is only shown while the settings overlay is toggling.
      ActionBar a = getSupportActionBar();
      if (a != null) {
        a.show();
      }
    } else {
      controller.hide(WindowInsetsCompat.Type.systemBars());
      controller.setSystemBarsBehavior(
          WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE);
    }

    HideRunnable r = getHideRunnable();
    if (r != null) {
      mainView.removeCallbacks(r);
      mainView.postDelayed(r, 3000);
    }
  }

  private void toast(final String text) {
    runOnUiThread(() -> Toast.makeText(MainActivity.this, text, Toast.LENGTH_LONG).show());
  }

  public SenderService getSenderService() {
    return mSender;
  }

  public void setSenderService(SenderService sender) {
    this.mSender = sender;
  }

  private HideRunnable getHideRunnable() {
    return mHideRunnable;
  }

  private void setHideRunnable(HideRunnable r) {
    this.mHideRunnable = r;
  }

  private class HideRunnable implements Runnable {

    @Override
    public void run() {
      setSystemUiVisibility(false);
    }
  }

  @Override
  protected void onDestroy() {
    mSensorManager.unregisterListener(this);

    if (mVideo != null) {
      mVideo.stopPlayback();
      mVideo.setOnStreamErrorListener(null);
      mVideo = null;
    }
    final View mainView = findViewById(R.id.main);
    mainView.removeCallbacks(getHideRunnable());
    final ViewGroup buttonsView = findViewById(R.id.buttons);
    if (buttonsView != null) {
      for (int i = 0; i < buttonsView.getChildCount(); ++i) {
        buttonsView.getChildAt(i).setOnClickListener(null);
      }
    }

    final Button btnSettings = findViewById(R.id.btnSettings);
    if (btnSettings != null) {
      btnSettings.setOnClickListener(null);
    }

    final SquareTouchPadLayout pad1 = findViewById(R.id.leftPad);
    if (pad1 != null) {
      pad1.setSender(null);
    }

    final SquareTouchPadLayout pad2 = findViewById(R.id.rightPad);
    if (pad2 != null) {
      pad2.setSender(null);
    }

    PreferenceManager.getDefaultSharedPreferences(getBaseContext())
        .unregisterOnSharedPreferenceChangeListener(mSharedPreferencesListener);
    getSenderService().setOnDisconnectedListener(null);
    getSenderService().setShowTextCallback(null);
    setSenderService(null);
    setHideRunnable(null);
    super.onDestroy();
  }
}

File: app/src/main/kotlin/com/trikset/gamepad/MainActivity.kt
package com.trikset.gamepad

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import android.content.pm.ActivityInfo
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.os.Bundle
import android.view.KeyEvent
import android.view.Menu
import android.view.MenuItem
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.view.animation.AlphaAnimation
import android.widget.Button
import android.widget.Toast
import androidx.activity.viewModels
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.WindowCompat
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.preference.PreferenceManager
import com.trikset.gamepad.diagnostics.AppLog
import com.trikset.gamepad.diagnostics.CrashLogStore
import com.trikset.gamepad.diagnostics.DiagnosticsReport
import com.trikset.gamepad.diagnostics.ReportSharer
import com.trikset.gamepad.mjpeg.MjpegView
import java.net.URL
import kotlinx.coroutines.launch

/**
 * The gamepad: two touch pads, five magic buttons, a sensor-driven wheel and the MJPEG video
 * stream, all sending commands through [SenderService].
 */
class MainActivity :
    AppCompatActivity(), SensorEventListener, MainActivitySettingsController.SettingsUi {

  private var mSensorManager: SensorManager? = null
  private var mAngle = 0 // -100% .. +100%
  private var mWheelEnabled = false
  private var mWheelStep = WHEEL_STEP_DEFAULT
  private var mVideo: MjpegView? = null
  private var mVideoURL: URL? = null
  private var mSettingsController: MainActivitySettingsController? = null
  private var videoRetryController: VideoRetryController? = null
  private val senderViewModel: SenderViewModel by viewModels()
  private val wheelController = WheelController()
  private val magicButtons = MagicButtonPanel(this) { getSenderService().send(it) }
  // Lazy: `window` is only assigned during Activity.attach(), which runs after
  // construction — a field initializer touching it would NPE/throw in Robolectric.
  private val systemUiController: SystemUiController by lazy {
    SystemUiController(
        window,
        mainViewProvider = { findViewById(R.id.main) },
        actionBarProvider = { supportActionBar },
        hideDelayMs = HIDE_DELAY_MS,
    )
  }
  private val connectionFeedback: ConnectionFeedback by lazy {
    ConnectionFeedback(
        context = this,
        settingsButtonProvider = { findViewById(R.id.btnSettings) },
        rootViewProvider = { findViewById(R.id.main) },
        statusTextProvider = { findViewById(R.id.connectionStatus) },
        targetConfiguredProvider = { !getSenderService().getHostAddr().isNullOrBlank() },
        connectAction = { getSenderService().connect() },
    )
  }
  private val hardwareGamepadController: HardwareGamepadController by lazy {
    HardwareGamepadController(
        send = { getSenderService().send(it) },
        settings = {
          val prefs = PreferenceManager.getDefaultSharedPreferences(this)
          HardwareGamepadController.Settings(
              swapSticks = prefs.getBoolean(SettingsFragment.SK_GAMEPAD_SWAP, false),
              magicButtonCount = MainActivitySettingsController.readMagicButtonCount(prefs),
          )
        },
    )
  }
  private val videoStreamErrorNotifier = VideoStreamErrorNotifier()

  private fun createPad(id: Int, strId: String) {
    val pad = findViewById<SquareTouchPadLayout>(id)
    if (pad != null) {
      pad.setPadName("pad $strId")
      pad.setSender(getSenderService())
    }
  }

  override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {
    // no-op, kept for the SensorEventListener contract
  }

  override fun onCreate(savedInstanceState: Bundle?) {
    requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE
    // Edge-to-edge (required on API 35+; enforced for targetSdk 36): draw
    // behind the system bars instead of using the removed FLAG_FULLSCREEN.
    WindowCompat.setDecorFitsSystemWindows(window, false)
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_main)
    systemUiController.setVisibility(false)
    connectionFeedback.attach()
    val actionBar = supportActionBar
    if (actionBar != null) {
      actionBar.setDisplayShowHomeEnabled(true)
      actionBar.setDisplayUseLogoEnabled(false)
      actionBar.setLogo(R.mipmap.trik_gamepad_logo_512x512)
      actionBar.setDisplayShowTitleEnabled(true)
    }

    mSensorManager = getSystemService(Context.SENSOR_SERVICE) as SensorManager

    mVideo = findViewById(R.id.video)

    getSenderService().setShowTextCallback { message ->
      // The gear border is the persistent status; surface only connection *errors* as feedback,
      // using a Snackbar (Material) instead of a transient Toast.
      if (message.endsWith(ERROR_SUFFIX)) {
        runOnUiThread { connectionFeedback.error(message) }
      }
    }

    val btnSettings = findViewById<Button>(R.id.btnSettings)
    if (btnSettings != null) {
      btnSettings.setOnClickListener {
        val actionBar1 = supportActionBar
        if (actionBar1 != null) {
          systemUiController.setVisibility(!actionBar1.isShowing)
        }
      }
    }

    val controlsOverlay = findViewById<View>(R.id.controlsOverlay)
    if (controlsOverlay != null) {
      controlsOverlay.bringToFront()
    }

    createPad(R.id.leftPad, "1")
    createPad(R.id.rightPad, "2")

    mSettingsController = MainActivitySettingsController(this, getSenderService(), this)
    mSettingsController?.register()

    // Bounded video-stream retry, gated on the control connection's keepalive (connectionState is
    // Connected == the same robot is reachable). The gate relaxes for an empty host: a video-only
    // device (no control target) must still auto-recover its stream, so retries run on the URL +
    // not-playing checks alone there. Reloads only while the view is not playing, so a healthy
    // stream is never disturbed; a failed open / silent stall / foldable surface recreation all
    // leave the view not-playing and are recovered by the 5 s tick or by the control-Connected
    // edge (see VideoRetryController).
    videoRetryController =
        VideoRetryController(
            shouldReload = { shouldReloadVideo() },
            reload = { restartVideoStream() },
        )

    // Observe the TCP connection state for the activity's lifetime; repeatOnLifecycle
    // (not the deprecated launchWhenX) stops the collection on STOP and restarts it
    // fresh on each START, so no stale emissions are collected across pauses.
    lifecycleScope.launch {
      repeatOnLifecycle(Lifecycle.State.STARTED) {
        senderViewModel.connectionState.collect { state ->
          connectionFeedback.update(state)
          if (
              state is ConnectionState.Disconnected &&
                  state.reason.isNotEmpty() &&
                  state.reason != PAUSE_DISCONNECT_REASON
          ) {
            connectionFeedback.error("Disconnected." + state.reason)
          }
          if (state is ConnectionState.Connected) {
            // Edge trigger: robot reachable again -> reload the video right away if it is dead.
            videoRetryController?.onControlConnected()
          }
        }
      }
    }

    maybeShowCrashReportDialog()
  }

  /**
   * Surfaces a captured crash exactly once: offers to share the report via the text-editor flow (or
   * the direct share sheet when "Share without editing" is set), copy it, or dismiss. The report
   * carries the live connection state and current log tail next to the crash trace.
   */
  private fun maybeShowCrashReportDialog() {
    val store = CrashLogStore(this)
    if (!store.shouldPrompt()) {
      return
    }
    store.markPrompted()
    val crash = store.latest() ?: return
    val shareWithoutEditing =
        PreferenceManager.getDefaultSharedPreferences(this)
            .getBoolean(SettingsFragment.SK_SHARE_WITHOUT_EDITING, false)
    val reportText = buildCrashReport(store, crash.stackTrace)
    AlertDialog.Builder(this)
        .setTitle(R.string.crash_dialog_title)
        .setMessage(R.string.crash_dialog_message)
        .setPositiveButton(
            if (shareWithoutEditing) R.string.share else R.string.review_and_share
        ) { _, _ ->
          ReportSharer.share(this, reportText)
        }
        .setNeutralButton(R.string.copy) { _, _ -> copyReport(reportText) }
        .setNegativeButton(R.string.dismiss, null)
        .show()
  }

  private fun buildCrashReport(store: CrashLogStore, crashTrace: String): String {
    val prefs = PreferenceManager.getDefaultSharedPreferences(this)
    return DiagnosticsReport.build(
        this,
        prefs,
        senderViewModel.connectionState.value,
        AppLog.tail(AppLog.BUFFER_CAPACITY),
        crashTrace,
    )
  }

  private fun copyReport(reportText: String) {
    val clipboard = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
    clipboard.setPrimaryClip(ClipData.newPlainText(getString(R.string.report_subject), reportText))
    Toast.makeText(this, R.string.copied_to_clipboard, Toast.LENGTH_SHORT).show()
  }

  override fun onCreateOptionsMenu(menu: Menu): Boolean {
    menuInflater.inflate(R.menu.menu, menu)
    return true
  }

  override fun onOptionsItemSelected(item: MenuItem): Boolean =
      when (item.itemId) {
        R.id.settings -> {
          startActivity(Intent(this, SettingsActivity::class.java))
          true
        }
        else -> super.onOptionsItemSelected(item)
      }

  override fun dispatchKeyEvent(event: KeyEvent): Boolean {
    // A hardware gamepad maps to pad/button commands; unmapped keys fall through.
    val handled =
        when (event.action) {
          KeyEvent.ACTION_DOWN ->
              hardwareGamepadController.onKeyDown(event.keyCode, event.repeatCount)
          KeyEvent.ACTION_UP -> hardwareGamepadController.onKeyUp(event.keyCode)
          else -> false
        }
    return handled || super.dispatchKeyEvent(event)
  }

  override fun onGenericMotionEvent(event: MotionEvent): Boolean =
      hardwareGamepadController.onMotionEvent(event) || super.onGenericMotionEvent(event)

  override fun onPause() {
    mSensorManager?.unregisterListener(this)
    getSenderService().disconnect(PAUSE_DISCONNECT_REASON)
    videoRetryController?.onPause()
    setVideoLoading(false)
    val video = mVideo
    if (video != null) {
      video.stopPlayback()
      video.setOnStreamErrorListener(null)
      video.setOnFirstFrameListener(null)
    }
    super.onPause()
  }

  override fun onResume() {
    super.onResume()
    videoRetryController?.onResume()
    val video = mVideo
    if (video != null) {
      // Reconnect-on-error: the render thread reports a dead stream and we
      // drop the HTTP connection and restart it (R12; see DECISIONS.md
      // "MJPEG: reconnect-on-error").
      video.setOnStreamErrorListener { videoRetryController?.onStreamError() }
      // Hide the loading indicator once the first frame of this playback cycle renders.
      video.setOnFirstFrameListener { runOnUiThread { setVideoLoading(false) } }
      restartVideoStream()
    }
    val sensorManager = mSensorManager
    if (sensorManager != null) {
      sensorManager.registerListener(
          this,
          sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER),
          SensorManager.SENSOR_DELAY_NORMAL,
      )
    }
  }

  /**
   * Retry gate for the bounded video-stream reload. Requires a configured URL and a not-playing
   * view; the control connection must be Connected, unless the host is blank — a video-only device
   * (no control target) still needs its stream to auto-recover.
   */
  private fun shouldReloadVideo(): Boolean {
    val hostConfigured = !getSenderService().getHostAddr().isNullOrBlank()
    return (!hostConfigured ||
        senderViewModel.connectionState.value is ConnectionState.Connected) &&
        mVideoURL != null &&
        mVideo?.isPlaying() == false
  }

  private fun restartVideoStream() {
    // The error listener may fire from the render thread; always hop to the
    // main thread before touching the view hierarchy / opening the stream.
    runOnUiThread {
      val video = mVideo ?: return@runOnUiThread
      // Show the loading indicator only while a URL is configured AND the control connection is
      // up (no spinner when disconnected / no stream URL); it stays up until the first frame
      // renders (robot video disabled -> keeps cycling).
      if (mVideoURL != null && senderViewModel.connectionState.value is ConnectionState.Connected) {
        setVideoLoading(true)
      }
      // Feed the load outcome back into the retry controller: a failed open arms the bounded
      // retry loop, a success disarms it. A failure also surfaces a throttled "video stream
      // unavailable" Snackbar (once per failure episode, not per 5 s retry tick).
      VideoStreamLoader(video).load(mVideoURL) { ok ->
        if (ok) {
          videoRetryController?.onLoadSuccess()
        } else {
          videoRetryController?.onLoadFailed()
          if (videoStreamErrorNotifier.shouldNotify()) {
            connectionFeedback.error(getString(R.string.video_stream_unavailable))
          }
        }
      }
    }
  }

  private fun setVideoLoading(visible: Boolean) {
    findViewById<android.widget.ProgressBar>(R.id.videoLoading)?.visibility =
        if (visible) View.VISIBLE else View.GONE
  }

  override fun onSensorChanged(event: SensorEvent) {
    if (event.sensor.type == Sensor.TYPE_ACCELEROMETER) {
      processSensor(event.values)
    }
  }

  private fun processSensor(values: FloatArray) {
    val angle =
        wheelController.nextAngle(values[0], values[1], mAngle, mWheelStep, mWheelEnabled) ?: return
    mAngle = angle
    getSenderService().send("wheel $mAngle")
  }

  override fun setActionBarTitle(title: String): Boolean {
    val actionBar = supportActionBar
    return if (actionBar != null) {
      actionBar.title = title
      true
    } else {
      false
    }
  }

  override fun toast(text: String) {
    runOnUiThread { Toast.makeText(this, text, Toast.LENGTH_LONG).show() }
  }

  override fun animatePadsAlpha(alpha: Float, previousAlpha: Float) {
    val alphaUp = AlphaAnimation(previousAlpha, alpha)
    alphaUp.setFillAfter(true)
    alphaUp.setDuration(ALPHA_ANIMATION_MS)
    findViewById<View>(R.id.controlsOverlay)?.startAnimation(alphaUp)
    findViewById<View>(R.id.buttons)?.startAnimation(alphaUp)
  }

  override fun setVideoUrl(url: URL?) {
    mVideoURL = url
    // No URL -> show a hint instead of a silent black area. The placeholder stays hidden while a
    // URL is configured, even if the stream is down (the loading spinner + gear border convey that
    // state).
    findViewById<android.widget.TextView>(R.id.videoPlaceholder)?.visibility =
        if (url == null) View.VISIBLE else View.GONE
  }

  override fun setKeepScreenOn(enabled: Boolean) {
    findViewById<View>(R.id.main)?.keepScreenOn = enabled
  }

  override fun setMagicButtons(count: Int, symbols: List<String>) {
    val buttonsView = findViewById<ViewGroup>(R.id.buttons) ?: return
    magicButtons.populate(buttonsView, count, symbols)
  }

  override fun setControlsVisible(visible: Boolean) {
    // GONE removes the pads/buttons from the layout entirely (no hit-testing).
    val visibility = if (visible) View.VISIBLE else View.GONE
    findViewById<View>(R.id.controlsOverlay)?.visibility = visibility
    findViewById<View>(R.id.buttons)?.visibility = visibility
  }

  override fun setShowFps(enabled: Boolean) {
    mVideo?.showFps = enabled
  }

  override fun getWheelStep(): Int = mWheelStep

  override fun setWheelStep(step: Int) {
    mWheelStep = step
  }

  override fun isWheelEnabled(): Boolean = mWheelEnabled

  override fun setWheelEnabled(enabled: Boolean) {
    mWheelEnabled = enabled
  }

  fun getSenderService(): SenderService = senderViewModel.sender

  fun getSettingsController(): MainActivitySettingsController? = mSettingsController

  fun setSenderService(sender: SenderService?) {
    if (sender != null) {
      senderViewModel.sender = sender
    }
  }

  override fun onDestroy() {
    mSensorManager?.unregisterListener(this)
    setVideoLoading(false)
    val video = mVideo
    if (video != null) {
      video.stopPlayback()
      video.setOnStreamErrorListener(null)
      video.setOnFirstFrameListener(null)
      mVideo = null
    }
    systemUiController.detach()
    val buttonsView = findViewById<ViewGroup>(R.id.buttons)
    if (buttonsView != null) {
      magicButtons.clearListeners(buttonsView)
    }
    findViewById<Button>(R.id.btnSettings)?.setOnClickListener(null)
    findViewById<SquareTouchPadLayout>(R.id.leftPad)?.setSender(null)
    findViewById<SquareTouchPadLayout>(R.id.rightPad)?.setSender(null)
    mSettingsController?.unregister()
    mSettingsController = null
    getSenderService().setShowTextCallback(null)
    super.onDestroy()
  }

  private companion object {
    const val TAG = "MainActivity"
    const val HIDE_DELAY_MS = 3000L
    const val ALPHA_ANIMATION_MS = 2000L
    const val WHEEL_STEP_DEFAULT = 7
    const val PAUSE_DISCONNECT_REASON = "Inactive gamepad"
    const val ERROR_SUFFIX = " error."
  }
}
